### ISSUE TITLE
Logging source file and line via SPDLOG_LOGGER_CALL

### ISSUE THREAD

**Issue by caibf** [gabime/spdlog#1582]
Hi, 

I have set spdlog's logging format with the following pattern:
`spdlog::set_pattern("[%Y-%m-%d %X.%e] [%^%l%$] [%s:%#] [%!] - %v");`

It would format log messages in the form of:
`[2014-10-31 23:46:59.678] [info] [my_file.cpp:123] [my_func] - Some message`

Main messages I want to log were source file name and line number where I logged. To reduce duplicate code, I made the log function as the member function of a class.
```
template<typename T>
void FuncUtility::spdLogStatic(const T &msg, spdlog::level::level_enum lvl)
{
	try
	{
		/*_daily_logger->log(lvl, msg);
		_console_logger->log(lvl, msg);*/

		SPDLOG_LOGGER_CALL(_daily_logger, lvl, msg);
		SPDLOG_LOGGER_CALL(_console_logger, lvl, msg);
	}
	catch (const spdlog::spdlog_ex &ex)
	{
		std::printf("Log failed: %s\n", ex.what());
	}
}
```

Then, others could log something in anywhere simply via:
`FuncUtility::spdLogStatic("This is a testing message.");`

However, source file name and line number of the logger always logged _FuncUtility::spdLogStatic_ 
and its line number. That was not I expect. How could I do that?

**Comment by tt4g** [comment]
spdlog gets source file information from `spdlog::source_loc`,  and that is provided by `SPDLOG_LOGGER_CALL` marco.

`SPDLOG_LOGGER_CALL` specifies `__LINE__`, `__FILE__` and `SPDLOG_FUNCTION` macros as arguments for `spdlog::source_loc` constructor.

https://github.com/gabime/spdlog/blob/22a169bc319ac06948e7ee0be6b9b0ac81386604/include/spdlog/spdlog.h#L286

Should accept `spdlog::source_loc` and define custom macro.

Example:

```diff
template<typename T>
- void FuncUtility::spdLogStatic(const T &msg, spdlog::level::level_enum lvl)
+ void FuncUtility::spdLogStatic(spdlog::source_loc source, const T &msg, spdlog::level::level_enum lvl)
{
	try
	{
		/*_daily_logger->log(lvl, msg);
		_console_logger->log(lvl, msg);*/

-		SPDLOG_LOGGER_CALL(_daily_logger, lvl, msg);
-		SPDLOG_LOGGER_CALL(_console_logger, lvl, msg);
+             _daily_logger->log(source, lvl, msg);
+             _console_logger->log(source, lvl, msg);
	}
	catch (const spdlog::spdlog_ex &ex)
	{
		std::printf("Log failed: %s\n", ex.what());
	}
}

+ #define MY_OWN_LOGGER_CALL(msg, lvl) FuncUtility::spdLogStatic(spdlog::source_loc{__FILE__, __LINE__, SPDLOG_FUNCTION}, msg, lvl);

**Comment by caibf** [comment]
Great help! Thanks @tt4g 

By follow your instructions, I have actually logged the source file name and the corresponding line number.

Here attaching the codeds for others who face the same issue with me.
```
#ifndef FUNC_UTILITY_IMPL_H
#define FUNC_UTILITY_IMPL_H

// https://github.com/gabime/spdlog/issues/1582
#define SGRT_LOGGER_CALL(msg, lvl) FuncUtility::spdLog(spdlog::source_loc{__FILE__, __LINE__, SPDLOG_FUNCTION}, msg, lvl)
#define SGRT_LOGGER_DEBUG(msg) SGRT_LOGGER_CALL(msg, spdlog::level::debug)
#define SGRT_LOGGER_INFO(msg) SGRT_LOGGER_CALL(msg, spdlog::level::info)
#define SGRT_LOGGER_ERROR(msg) SGRT_LOGGER_CALL(msg, spdlog::level::err)

template<typename T>
void FuncUtility::spdLog(spdlog::source_loc source, const T &msg, spdlog::level::level_enum lvl)
{
	try
	{
		/*_daily_logger->log(lvl, msg);
		_console_logger->log(lvl, msg);*/

		_daily_logger->log(source, lvl, msg);
		_console_logger->log(source, lvl, msg);
	}
	catch (const spdlog::spdlog_ex &ex)
	{
		std::printf("Log failed: %s\n", ex.what());
	}
}

#endif
```

**Comment by caibf** [comment]
Nowadays, I want to use the library to log something from C# application layer through P/Invoke technology. 

Code snippets in C++ as follows:
```
void
FuncUtility::spdLogInterop(const char *filename_in, int line_in, const char *funcname_in, const char* msg, spdlog::level::level_enum lvl)
{
	char* tmp_filename = (char*)CoTaskMemAlloc(strlen(filename_in) + 1);
	char* tmp_funcname = (char*)CoTaskMemAlloc(strlen(funcname_in) + 1);
	char* tmp_msg = (char*)CoTaskMemAlloc(strlen(msg) + 1);

	strcpy(tmp_filename, filename_in);
	strcpy(tmp_funcname, funcname_in);
	strcpy(tmp_msg, msg);

	_daily_logger->log(spdlog::source_loc{ tmp_filename, line_in, static_cast<const char *>(tmp_funcname) }, lvl, tmp_msg);

	CoTaskMemFree(tmp_filename);
	CoTaskMemFree(tmp_funcname);
	CoTaskMemFree(tmp_msg);
}
```

The declaration of caller in C# like this:
```
[DllImport("EnsensoInterop.dll")]
public static extern void LoggerInterop(string filename_in, int line_in, string funcname_in, string msg, LogLevel lvl);
```

However, when I called this interop function from C#, broken message occurs.
![image](https://user-images.githubusercontent.com/4711529/89140284-c61d8480-d573-11ea-8f79-879615c33f50.png)


How do I fix it? @tt4g

**Comment by tt4g** [comment]
Should you add keywrods `extern "C"` and `__declspec(dllexport)` to `FuncUtility::spdLogInterop()`.

For more information on exporting functions from C++, you may find the following links helpful: https://stackoverflow.com/questions/14395192/why-does-c11-not-support-declaring-extern-c-on-a-static-member-function

**Comment by tt4g** [comment]
I'm not familiar with C#, but I think the C# function declarations are also wrong.
To pass C# `string` to C/C++ `char *`, you need to add [MarshalAs(UnmanagedType.LPStr)] keyword before `string`.

See official document: https://docs.microsoft.com/en-us/dotnet/framework/interop/default-marshaling-for-strings

Example:

```csharp
[DllImport("EnsensoInterop.dll")]
public static extern void LoggerInterop(
        [MarshalAs(UnmanagedType.LPStr)] string filename_in, 
        int line_in,
        [MarshalAs(UnmanagedType.LPStr)] string funcname_in,
        [MarshalAs(UnmanagedType.LPStr)] string msg, 
        LogLevel lvl);
```

**Comment by caibf** [comment]
`FuncUtility::spdLogInterop()` is not the export function, it was called from the export function that defined in other files.
```
extern "C" __declspec(dllexport) void 
LoggerInterop(const char *filename_in, int line_in, const char *funcname_in, const char* msg, spdlog::level::level_enum lvl)
{
	window->getFunc().spdLogInterop(filename_in, line_in, funcname_in, msg, lvl);
}
```

It's not the parameters passing problem because I used this way in other P/Invoke method. To make a clear testing, I used the `constant` string instead of string from C#. But it's the same problem
```
void
FuncUtility::spdLogInterop(const char *filename_in, int line_in, const char *funcname_in, const char* msg, spdlog::level::level_enum lvl)
{
	char* tmp_filename = (char*)CoTaskMemAlloc(strlen(filename_in) + 1);
	char* tmp_funcname = (char*)CoTaskMemAlloc(strlen(funcname_in) + 1);
	char* tmp_msg = (char*)CoTaskMemAlloc(strlen(msg) + 1);

	strcpy(tmp_filename, filename_in);
	strcpy(tmp_funcname, funcname_in);
	strcpy(tmp_msg, msg);

	_daily_logger->log(spdlog::source_loc{ __FILE__, __LINE__, SPDLOG_FUNCTION }, spdlog::level::debug, "Just a testing message");

	CoTaskMemFree(tmp_filename);
	CoTaskMemFree(tmp_funcname);
	CoTaskMemFree(tmp_msg);
}
```

**Comment by tt4g** [comment]
Does adding the `[MarshalAs(UnmanagedType.LPStr), In]` keyword not change the result?

P.S I forgot to mention that C# string seems to have the same character encoding as `wchar_t`.
The internal character encoding in this case seems to be UTF-16, so you may need to use `[MarshalAs(UnmanagedType.LPWStr), In]` instead of  `[MarshalAs(UnmanagedType.LPStr), In]`.
And in order for spdlog to handle UTF-16, you need to have CMake variable `SPDLOG_WCHAR_SUPPORT` defined at build time.

**Comment by caibf** [comment]
Hey, I think this is not the parameters passing problem because testing with constant string would lead to same result.
`_daily_logger->log(spdlog::source_loc{ __FILE__, __LINE__, SPDLOG_FUNCTION }, spdlog::level::debug, "Just a testing message");`

**Comment by tt4g** [comment]
When you pass a C# string to `CoTaskMemAlloc()` or `strlen()`, it may not be doing what you expect it to do.

Can you confirm that the C# string is a null terminated `char *` and that the `strlen()` gets the exact length of the character?

**Comment by caibf** [comment]
Dear @tt4g 

I have figured out what's the real culprit that causes this problem. I defined `_daily_logger` as the `static` in the `FuncUtility` class, then I make a trick to initialize the `static variable` via `static constructor`.
 ```
class FuncUtility
{
public:
    // ...
private:
    static std::shared_ptr<spdlog::logger> _daily_logger;
}

std::shared_ptr<spdlog::logger> FuncUtility::_daily_logger;
```

If remove the static declaration, no exception throws. However, I am still a green hand with C++, could you tell me what's wrong with it?

**Comment by tt4g** [comment]
A common problem is that if a static member variable of a class is declared in a library header file, that variable will be defined at a different address in the library and the application.
In this case, the application and the library are not accessing the same variables, so even though the library may have initialized the static variables, the static variables the application references may not have been initialized.

Since the header file's inclusion is a copy of the content, imagine that the static variable definitions exist in two places: the application and the library.

As far as I know, an easy workaround is to initialize the static variable in the static function and return it.

Example:

`FuncUtility.hpp`
```cpp

class FuncUtility
{
public:

    static std::shared_ptr<spdlog::logger> daily_logger(); // declare static function.
};
```

`FuncUtility.cpp`:
```cpp

// implementation.
std::shared_ptr<spdlog::logger> FuncUtility::daily_logger()
{
    static std::shared_ptr<spdlog::logger> logger_ = // Initialize spdlog::logger here!

    return logger_;
}
```

Using this technique, there are no static variables in the header file and your application will access the static variables managed by the library.

NOTE: In C++11 and later, the spec specifies that the initialization of static local variables must be thread-safe.

**Comment by caibf** [comment]
Wow! It's a amazing things that the static variable defined at a different address in the library and the application. I learned new knowledge from you again. Very thanks @tt4g 

I adopted this workaround and modified the code, it really no exception threw out. Furthermore, I commited the full code to my repository, someone can read the code if you are insteresting in this issue.

Repository link: [https://github.com/caibf/TestLog](https://github.com/caibf/TestLog)
